File: day16/permutation_promenade.py
#!/usr/bin/env python3
from sys import argv
import re

# parse the input
moves = []
filename = argv[1]
with open(filename) as f:
    for line in f:
        line = line.rstrip()
        moves = line.split(",")

pgm = [chr(x + ord('a')) for x in range(16)]

for move in moves:
    cmd = move[0]
    if cmd == 's':
        m = re.search('s(\d+)', move)
        pos = len(pgm) - int(m.group(1))
        pgm = pgm[pos:] + pgm[:pos]
    elif cmd == 'x':
        m = re.search('x(\d+)/(\d+)', move)
        pos1 = int(m.group(1))
        pos2 = int(m.group(2))
        tmp = pgm[pos1]
        pgm[pos1] = pgm[pos2]
        pgm[pos2] = tmp
    elif cmd == 'p':
        d = {v: i for (i,v) in enumerate(pgm)}
        tmp = pgm[d[move[1]]]
        pgm[d[move[1]]] = pgm[d[move[3]]]
        pgm[d[move[3]]] = tmp

print("result1:", "".join(pgm))

File: day16/permutation_promenade.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);

my @moves;
while (<>) {
    chomp;
    @moves = split /,/;
}

#my @pgm = 'a'..'e';
my @pgm = 'a'..'p';
for my $move (@moves) {
    my $cmd = substr $move, 0, 1;
    if ($cmd eq 's') {
        $move =~ /s(\d+)/;
        @pgm = (@pgm[-$1..-1], @pgm[0..$#pgm-$1]);
    } elsif ($cmd eq 'x') {
        $move =~ /x(\d+)\/(\d+)/;
        my $tmp = $pgm[$1];
        $pgm[$1] = $pgm[$2];
        $pgm[$2] = $tmp;
    } elsif ($cmd eq 'p') {
        my @tok = split //, $move;
        my %h = map { $pgm[$_] => $_ } 0..$#pgm;
        my $tmp = $pgm[$h{$tok[1]}];
        $pgm[$h{$tok[1]}] = $pgm[$h{$tok[3]}];
        $pgm[$h{$tok[3]}] = $tmp;
    }
}

say "result1: ", join "", @pgm;
